### DISCUSSION TITLE
One site will not work through Zoraxy

### DISCUSSION THREAD

**Discussion by ckvist72** [tobychui/zoraxy#228] [Q&A] [answered]
Hi.
I have moved all from Nginx to Zoraxy. And everything works. Except my proxu for Opnsense... It hust says bad request. - Any idea if i have missed something.

**Comment by tobychui** [comment]
@ckvist72 I never used Opnsense but from its [documentation for reverse proxy](https://docs.opnsense.org/manual/how-tos/nginx.html), seems you will need to change something in their system settings to make it works. Try the followings

1. Make sure your Opnsense is running under HTTP or HTTPS with a valid certificate
2. Check "Skip certificate validation" in Zoraxy if your Opnsense instance is using self sign certificate with https 
3. Check your settings in Opnsense to make sure it accept the hostname you are using as one of its up(down)stream

**Comment by turbo1781** [comment]
Hello, I'm currently facing the same problem. I've got all servers running except OPNsense. Is there a solution here by the time this post was written?

**Reply by tobychui** [reply]
No, the original author just disappeared after dropping this question. What issues / error you are seeing on your side?

**Reply by turbo1781** [reply]
Exactly the same problem.

1. Make sure your Opnsense is running under HTTP or HTTPS with a valid certificate - **I have already a certificate with the ACME Plugin in the OPNsense.**
2. Check "Skip certificate validation" in Zoraxy if your Opnsense instance is using self sign certificate with https - **is checked**
3. Check your settings in Opnsense to make sure it accept the hostname you are using as one of its up(down)stream - **is entered in "Alternate Hostnames"**

I didn't have to change anything special in the OPNsense config in the nginx proxy manager, so I'm surprised that it doesn't work with Zoraxy

**Reply by tobychui** [reply]
In fact it is another way round. Since nginx is much older than Zoraxy, many open source project use it (and Apache) as the compatibility test proxy when coding their web ui. I design Zoraxy to work in my homelab, so it might not fit everyone's needs. 

However, you can help to make Zoraxy supports Opnsense if you could maybe provide a bit more hints on what errors you encountered like network traces or wireshark captures.

**Reply by turbo1781** [reply]
I also think that it is the OPNsense problem, perhaps I have ticked a box incorrectly. When I click on the link in Zoraxy under HTTP Proxy, I am correctly redirected to the OPNsense Web GUI. If I click on the domain name via Zoraxy instead, a bad request page appears.

Here ist a screenshot, all other settings are default...
![Screenshot 2025-01-20 121145](https://github.com/user-attachments/assets/bd33164d-7845-4b5e-ae2e-3d9dfcb18b2b)

**Reply by tobychui** [reply]
I don't see any problem with your setup. You probably go create an issue on OPNsense's side and ask for their maintainers opinion on this.

**Reply by Monviech** [reply]
Maybe this sheds some light on it:
https://github.com/opnsense/plugins/issues/4471#issuecomment-2599639355

**Reply by tobychui** [reply]
> Maybe this sheds some light on it:
> https://github.com/opnsense/plugins/issues/4471#issuecomment-2599639355

Zoraxy only support http2. I guess this might not be the issue here. Though new ideas are welcomed :)

**Reply by FirebladeBMW** [reply]
I am currently facing the same issue. Is there any way to make OPNsense Webgui accessible via Zoraxy?

**Reply by tobychui** [reply]
@FirebladeBMW You tried disabling chunked transfer encoding?

**Reply by FirebladeBMW** [reply]
@tobychui  Yes, and all combinations of TLS checkboxes. It's either 400 Bad request or 404 not found.

<img width="589" height="325" alt="grafik" src="https://github.com/user-attachments/assets/1199f7bc-63da-4f04-98ce-1198622e5c1e" />


With this setting I am able to access the webgui by clicking on the IP address. 
Clicking on the domain name will result in an error.

Here are the possible settings on OPNsense site. (These settings work with NPM)

<img width="680" height="900" alt="grafik" src="https://github.com/user-attachments/assets/cb5cb01a-8598-4dee-9b5b-78fb701eb734" />

**Reply by CrazyWolf13** [reply]
@tobychui any news to this?

Is anything else needed, can we provide any log or so to get this working with zoraxy?
OPNsense is quite common in homelab.

Maybe you could also test this with a opnsense VM: https://docs.opnsense.org/manual/virtuals.html

**Reply by tobychui** [reply]
@CrazyWolf13 

As you might have notice I didn't use opnsense, so I will not be working on it unless one day I decided to deploy this in my homelab (btw this is how open source projects works in general, you only contribute to parts that interests you). If you got a fix for this, feel free to open a pull request so I will review and merge it into the Zoraxy project.

**Comment by CrazyWolf13** [comment]
Hi, after a long time digging into this, I finally found a workaround/fix!

Initial source thread on opnsense forums: https://forum.opnsense.org/index.php?topic=50038
Check here for the solution: https://forum.opnsense.org/index.php?topic=50038.msg255386#msg255386

To provide it in short here:
add this code:

> [!CAUTION]
> Use with caution, I'm no expert and don't fully know all the possible side effects this may have!

```bash
server.http-parseopts = ( "method-get-body" => "enable" )
```

To the file: `/usr/local/etc/inc/plugins.inc.d/webgui.inc` after line number 488, you can check here how the file looks to find the line with nano/vi: https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/webgui.inc#L488

> [!WARNING]  
> I'm unsure if this fix on opnsense is persistend across updates!

---

According to the opnsense mods, this is an issue on zoarxy side: https://forum.opnsense.org/index.php?topic=50038.msg255406#msg255406 which the moderator even posted in this thread already but was sadly dismissed by @tobychui https://github.com/tobychui/zoraxy/discussions/228#discussioncomment-12055893

**Reply by tobychui** [reply]
Hey @CrazyWolf13 thanks for finding a workaround!

I have take a look at the forum posts. I just want to point out a few things here. As quoted from the devs on the opensense forum: 

>This indicates Zoraxy sends HTTP GET requests with a body, which is very much discouraged. No browser sends GET requests with a payload. So that's an issue you might want to raise with the Zoraxy devs.

<img width="979" height="473" alt="圖片" src="https://github.com/user-attachments/assets/b35c8e00-929d-4cb6-8c42-4d3915df4a63" />

Here is a screenshot of RFC7231 page 24, both what Zoraxy is sending and what opensense web ui uses to serve HTTP are both correct implementation of HTTP protocol. 

<img width="807" height="367" alt="圖片" src="https://github.com/user-attachments/assets/f89296e5-d8c4-4d93-b79e-17ad96a0e60b" />

Here is another screenshot from RFC7230, which said except the HEAD request, all request type carries a body payload. 

> which the moderator even posted in this thread already but was sadly dismissed by @tobychui

I don't want to get involved into the debate on who is right or wrong here, but I guess you can refer to [here](https://github.com/golang/go/issues/53894#issuecomment-1197261482) for the reason why Zoraxy is behaving this way.

One more thing worth mentioning, the Zoraxy wiki is open for edit, so feel free to contribute a page over the wiki on how to apply the workaround on opensense web ui.

**Reply by CrazyWolf13** [reply]
Thanks for the explanation!

Sure, already working on it :D

**Reply by Monviech** [reply]
The workaround in OPNsense is not persistent and should not be documented as such. Since caddy and traefik can solve this via configuration and they both use the same go packets as Zoraxy, I would assume a working configuration can be configured in it (Zoraxy) as well, most likely by removing the alt-svc header or by setting the http protocol between the go library and the upstream server (lighttpd on opnsense) to http1.1.

Also it's not about blame or who is right or wrong, its ambigous since all implementations do the right thing in the end, its just a configuration issue as both http participants have different expectations.

https://github.com/opnsense/plugins/issues/4471#issuecomment-2742109624

**Reply by CrazyWolf13** [reply]
Just for reference, the workaround with alt-svc headers is not working for me.
As far as I know setting the HTTP version is not possible on zoraxy.

If it's not persistent it's probably better, if we have this as somewhat more official either from opnsense or zoraxy.

**Reply by CrazyWolf13** [reply]
@tobychui 
I've documented it here, as people in this thread will try it out anyway now; https://github.com/tobychui/zoraxy/wiki/OPNsense-behind-Zoraxy but ideally this would be fixed in zoraxy.

**Reply by tobychui** [reply]
@CrazyWolf13 I can try fix it by adding a force HTTP/1.1 option for proxy rules in Zoraxy, but since this is caused by compatibility issue of upstream dependencies, I am not sure if this is even possible to fix. 

Besides, I have nothing to test this on so I guess some help will be needed for testing it from the v3.3.1 dev branch.

**Reply by tobychui** [reply]
Anyone can try this commit to see if that works?
https://github.com/tobychui/zoraxy/commit/13d61130bb98c2cfdc337cb18ef6a417f622ad4c

The force HTTP/1.1 option is here
<img width="858" height="389" alt="圖片" src="https://github.com/user-attachments/assets/a0d32b76-cbb4-4a1b-9763-24968288358b" />

**Reply by CrazyWolf13** [reply]
@tobychui Thanks for the quick fix!

When visiting the page behind zoraxy I just get this: 
<img width="629" height="390" alt="grafik" src="https://github.com/user-attachments/assets/4931ffbd-ce83-4932-91aa-221b609fe3a3" />

and with curl I see this:

```bash
⚡tobia ❯❯ curl -vk https://opnsense.local.local/
* Host opnsense.local.local:443 was resolved.
* IPv6: (none)
* IPv4: 192.168.1.23
*   Trying 192.168.1.23:443...
* schannel: disabled automatic use of client certificate
* ALPN: curl offers http/1.1
* ALPN: server accepted http/1.1
* Established connection to opnsense.local.local (192.168.1.23 port 443) from 192.168.1.23 port 61575
* using HTTP/1.x
> GET / HTTP/1.1
> Host: opnsense.local.local
> User-Agent: curl/8.16.0
> Accept: */*
>
* schannel: remote party requests renegotiation
* schannel: renegotiating SSL/TLS connection
* schannel: SSL/TLS connection renegotiated
< HTTP/1.1 301 Moved Permanently
< Content-Length: 0
< Date: Mon, 22 Dec 2025 16:46:13 GMT
< Location: https://opnsense.local.local/
< Server: OPNsense
< Zr-Origin-Location: https://opnsense.local.local/
<
* Connection #0 to host opnsense.local.local:443 left intact
```

Here in very verbose:

<details>
  <summary>Spoiler</summary>

```bash
 ⚡tobia ❯❯ curl -vvvk https://opnsense.local.local/
17:47:33.225000 [0-x] == Info: [READ] client_reset, clear readers
17:47:33.254000 [0-0] == Info: Host opnsense.local.local:443 was resolved.
17:47:33.268000 [0-0] == Info: IPv6: (none)
17:47:33.276000 [0-0] == Info: IPv4: 192.168.1.23
17:47:33.284000 [0-0] == Info: [HTTPS-CONNECT] adding wanted h2
17:47:33.289000 [0-0] == Info: [HTTPS-CONNECT] added
17:47:33.294000 [0-0] == Info: [HTTPS-CONNECT] connect, init
17:47:33.303000 [0-0] == Info:   Trying 192.168.1.23:443...
17:47:33.314000 [0-0] == Info: [HTTPS-CONNECT] connect -> 0, done=0
17:47:33.321000 [0-0] == Info: [HTTPS-CONNECT] Curl_conn_connect(block=0) -> 0, done=0
17:47:33.336000 [0-0] == Info: [HTTPS-CONNECT] adjust_pollset -> 0, 1 socks
17:47:33.351000 [0-0] == Info: [SSL] cf_connect()
17:47:33.362000 [0-0] == Info: schannel: disabled automatic use of client certificate
17:47:33.381000 [0-0] == Info: ALPN: curl offers http/1.1
17:47:33.393000 [0-0] == Info: [SSL] cf_connect() -> 0, done=0
17:47:33.403000 [0-0] == Info: [HTTPS-CONNECT] connect -> 0, done=0
17:47:33.415000 [0-0] == Info: [HTTPS-CONNECT] Curl_conn_connect(block=0) -> 0, done=0
17:47:33.424000 [0-0] == Info: [SSL] adjust_pollset, POLLIN fd=484
17:47:33.431000 [0-0] == Info: [HTTPS-CONNECT] adjust_pollset -> 0, 1 socks
17:47:33.438000 [0-0] == Info: [SSL] cf_connect()
17:47:33.447000 [0-0] == Info: ALPN: server accepted http/1.1
17:47:33.456000 [0-0] == Info: [SSL] cf_connect() -> 0, done=1
17:47:33.465000 [0-0] == Info: [HTTPS-CONNECT] connect+handshake h2: 163ms, 1st data: 139ms
17:47:33.475000 [0-0] == Info: [HTTPS-CONNECT] connect -> 0, done=1
17:47:33.480000 [0-0] == Info: [HTTPS-CONNECT] Curl_conn_connect(block=0) -> 0, done=1
17:47:33.487000 [0-0] == Info: Established connection to opnsense.local.local (192.168.1.23 port 443) from 192.168.1.23 port 54123
17:47:33.499000 [0-0] == Info: [HTTPS-CONNECT] query ALPN
17:47:33.506000 [0-0] == Info: [SSL] query ALPN: returning 'http/1.1'
17:47:33.513000 [0-0] == Info: using HTTP/1.x
17:47:33.515000 [0-0] => Send header, 84 bytes (0x54)
0000: GET / HTTP/1.1
0010: Host: opnsense.local.local
002c: User-Agent: curl/8.16.0
0045: Accept: */*
0052:
17:47:33.529000 [0-0] == Info: schannel: remote party requests renegotiation
17:47:33.535000 [0-0] == Info: schannel: renegotiating SSL/TLS connection
17:47:33.541000 [0-0] == Info: schannel: SSL/TLS connection renegotiated
17:47:33.549000 [0-0] <= Recv header, 32 bytes (0x20)
0000: HTTP/1.1 301 Moved Permanently
17:47:33.554000 [0-0] == Info: [WRITE] [OUT] wrote 32 header bytes -> 32
17:47:33.565000 [0-0] == Info: [WRITE] [PAUSE] writing 32/32 bytes of type c -> 0
17:47:33.571000 [0-0] == Info: [WRITE] download_write header(type=c, blen=32) -> 0
17:47:33.580000 [0-0] == Info: [WRITE] client_write(type=c, len=32) -> 0
17:47:33.583000 [0-0] <= Recv header, 19 bytes (0x13)
0000: Content-Length: 0
17:47:33.587000 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=19) -> 0
17:47:33.591000 [0-0] == Info: [WRITE] [OUT] wrote 19 header bytes -> 19
17:47:33.595000 [0-0] == Info: [WRITE] [PAUSE] writing 19/19 bytes of type 4 -> 0
17:47:33.599000 [0-0] == Info: [WRITE] download_write header(type=4, blen=19) -> 0
17:47:33.604000 [0-0] == Info: [WRITE] client_write(type=4, len=19) -> 0
17:47:33.607000 [0-0] <= Recv header, 37 bytes (0x25)
0000: Date: Mon, 22 Dec 2025 16:47:34 GMT
17:47:33.612000 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=37) -> 0
17:47:33.616000 [0-0] == Info: [WRITE] [OUT] wrote 37 header bytes -> 37
17:47:33.621000 [0-0] == Info: [WRITE] [PAUSE] writing 37/37 bytes of type 4 -> 0
17:47:33.625000 [0-0] == Info: [WRITE] download_write header(type=4, blen=37) -> 0
17:47:33.630000 [0-0] == Info: [WRITE] client_write(type=4, len=37) -> 0
17:47:33.634000 [0-0] <= Recv header, 41 bytes (0x29)
0000: Location: https://opnsense.local.local/
17:47:33.639000 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=41) -> 0
17:47:33.645000 [0-0] == Info: [WRITE] [OUT] wrote 41 header bytes -> 41
17:47:33.648000 [0-0] == Info: [WRITE] [PAUSE] writing 41/41 bytes of type 4 -> 0
17:47:33.653000 [0-0] == Info: [WRITE] download_write header(type=4, blen=41) -> 0
17:47:33.657000 [0-0] == Info: [WRITE] client_write(type=4, len=41) -> 0
17:47:33.662000 [0-0] <= Recv header, 18 bytes (0x12)
0000: Server: OPNsense
17:47:33.666000 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=18) -> 0
17:47:33.670000 [0-0] == Info: [WRITE] [OUT] wrote 18 header bytes -> 18
17:47:33.673000 [0-0] == Info: [WRITE] [PAUSE] writing 18/18 bytes of type 4 -> 0
17:47:33.679000 [0-0] == Info: [WRITE] download_write header(type=4, blen=18) -> 0
17:47:33.684000 [0-0] == Info: [WRITE] client_write(type=4, len=18) -> 0
17:47:33.687000 [0-0] <= Recv header, 51 bytes (0x33)
0000: Zr-Origin-Location: https://opnsense.local.local/
17:47:33.692000 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=51) -> 0
17:47:33.697000 [0-0] == Info: [WRITE] [OUT] wrote 51 header bytes -> 51
17:47:33.701000 [0-0] == Info: [WRITE] [PAUSE] writing 51/51 bytes of type 4 -> 0
17:47:33.706000 [0-0] == Info: [WRITE] download_write header(type=4, blen=51) -> 0
17:47:33.712000 [0-0] == Info: [WRITE] client_write(type=4, len=51) -> 0
17:47:33.717000 [0-0] <= Recv header, 2 bytes (0x2)
0000:
17:47:33.719000 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=2) -> 0
17:47:33.723000 [0-0] == Info: [WRITE] [OUT] wrote 2 header bytes -> 2
17:47:33.728000 [0-0] == Info: [WRITE] [PAUSE] writing 2/2 bytes of type 4 -> 0
17:47:33.732000 [0-0] == Info: [WRITE] download_write header(type=4, blen=2) -> 0
17:47:33.738000 [0-0] == Info: [WRITE] client_write(type=4, len=2) -> 0
17:47:33.741000 [0-0] == Info: [WRITE] xfer_write_resp(len=200, eos=0) -> 0
17:47:33.747000 [0-0] == Info: [WRITE] [OUT] done
17:47:33.749000 [0-0] == Info: [READ] client_reset, clear readers
17:47:33.752000 [0-0] == Info: Connection #0 to host opnsense.local.local:443 left intact
```

</details>

Test setup was zoraxy running on my laptop with a dns rewrite opnsense.local.local to the IP of my laptop, then generating a selfsigned cert.

**Reply by Monviech** [reply]
@CrazyWolf13 Can you proxy to the backend without TLS (set the opnsense to HTTP)? The logs look like lighttpd refuses to establish the connection since it cannot determine if it was indeed TLS and thus forces a redirect to HTTPS which creates a loop. Using TLS termination and proxying via HTTP could prove this.

**Reply by tobychui** [reply]
@CrazyWolf13 Try manually adding `X-Forwarded-Proto: https` header to the downstream to upstream direction and see if how opnsense web server reacts? Seems opnsense wants to redirect the user to `https` without knowing that the user is already accessing the page in `https`. 

**Update**
I just double checked, seems Zoraxy already have that in the header before making the http roundtrip request. Then I guess I have no idea about this. Let me know if you find anything new.

**Reply by CrazyWolf13** [reply]
@tobychui @Monviech switching OPNsense to HTTP then proxied through zoraxy works!

When visiting if OPNsense is in HTTPS mode, I get this on a browser now:

<img width="1844" height="954" alt="grafik" src="https://github.com/user-attachments/assets/10be4e2c-dffd-4c50-b4ef-9a57e32c7a8b" />

While I can't replicate the infinite loop currently, yet I've not changed really anything despite restarting the zoraxy bin.

**Reply by Monviech** [reply, marked as answer]
@tobychui Thank you for implementing the http1.1 mode. Together with a simple switch to http on the opnsense (which is fine in networks in your own control) the usecase is now cleanly solved in my opinion.

If TLS to downstream should also be supported, its most likely some identity ambiguity going on (which happens very early in a request before headers are even accepted and evaluated by a downstream webserver). http2 largely avoids this by carrying scheme and authority explicitly, while http1.1 might require additional handling on the proxy side.
